### ISSUE TITLE
endian-ness flag not save in file for float data types

### ISSUE THREAD

**Issue by jswhit** [Unidata/netcdf-c#112]
Here's an example using netcdf4-python:

    import numpy as np
    import netCDF4 as nc4
    dtype = '>i8'
    xs = np.arange(10, dtype=dtype)
    nc = nc4.Dataset('./test.nc', mode='w')
    d = nc.createDimension('x', size=xs.size)
    v = nc.createVariable('x', xs.dtype, ('x'),
                              endian='big')
    v[:] = xs
    print v.endian()
    nc.close()
    nc = nc4.Dataset('./test.nc')
    v = nc.variables['x']
    print v.endian()
    nc.close()

running this script yields:

    big
    big

and `ncdump -s test.nc` shows

    netcdf test {
    dimensions:
	    x = 10 ;
    variables:
	    int64 x(x) ;
		    x:_Storage = "contiguous" ;
		    x:_Endianness = "big" ;

which is as expected.  But, if you change dtype='>f8' in the test script I get

    big
    native

and 'ncdump -s' shows that the `_Endianness` attribute is not set.  

So, it looks like for float variables the endian-ness attribute is not being saved in the file, despite the fact the nc_def_var_endian  is called.  This is not true for integer variables.

I could write a simple C program to demonstrate this, but it was quicker just to show it with python calling C.

Note:  I am running this on a little-endian machine running OS X.

**Comment by jswhit** [comment]
Here's a C program that illustrates this:

    #include <string.h>
    #include <netcdf.h>
    #define FILE_NAME "test.nc"
    #define NDIM 10
    #define NLON 20
    #define DIM_NAME "x"
    #define VAR_NAME "fv"
    #define VAR_NAME2 "iv"
    int
    main()
    {
      int ncid, dimid, varid, varid2, retval;
      retval = nc_create(FILE_NAME, NC_NETCDF4 | NC_CLOBBER, &ncid);
      retval = nc_def_dim(ncid, DIM_NAME, NDIM, &dimid);
      retval = nc_def_var(ncid, VAR_NAME, NC_FLOAT, 1, &dimid, &varid);
      retval = nc_def_var_endian(ncid, varid, NC_ENDIAN_BIG);
      retval = nc_def_var(ncid, VAR_NAME2, NC_INT, 1, &dimid, &varid2);
      retval = nc_def_var_endian(ncid, varid2, NC_ENDIAN_BIG);
      retval = nc_close(ncid);
      return 0;
    }

which produces

    netcdf test {
    dimensions:
    	x = 10 ;
    variables:
	    float fv(x) ;
		fv:_Storage = "contiguous" ;         <--  endian-ness not set for float variable
	    int iv(x) ;
	  	iv:_Storage = "contiguous" ;
		iv:_Endianness = "big" ;                <---  endian-ness set for integer variable

    // global attributes:
		    :_Format = "netCDF-4" ;

**Comment by jswhit** [comment]
Been 21 days, anyone have a chance to look at this yet?

Sure looks like a bug in the C library to me.

**Comment by WardF** [comment]
I've created a ticket in the Unidata JIRA system, viewable at https://bugtracking.unidata.ucar.edu/browse/NCF-331.  I'll keep it, and this entry on GitHub, updated.

**Comment by WardF** [comment]
So I've traced this down and it appears to be an issue in the hdf5 library.  I'm not completely certain of that, however, so I'm working on a pure hdf5 version of this test so that I can be certain.

**Comment by WardF** [comment]
I think we've got the issue addressed; it appears to have originated in `nc4file.c:get_type_info2()`. The call to `H5Tget_order()` was only happening for `H5T_Float`. According to the documentation, this works for all datatypes as of hdf5 1.8.6.

The good news is that the endianness flag **was** being saved to file; I was able to confirm this by writing out the file using `libhdf5` directly; `h5dump` would report the endianness properly, but ncdump would not.  

The fix has been committed in branch `ncf331`; I'm cleaning up a bit and then will merge into `master`.  I will close out the issue once this happens, hopefully today.
